### PR TITLE
feat(Communities): allow for creating community permissions

### PR DESCRIPTION
**Pull request by 0x-r4bbit**
This commit is the first of implementing community permissions. **It is not implementing the complete feature**, rather does it introduce the first pieces, such that we can get code reviewed and merged before it grows too big.

To review these features, please make sure to

1. Enable wallet (Settings -> Advanced -> Wallet toggle)
2. Enable community permissions (Settings -> Advanced -> Community Permissions toggle)

You'll have to restart the app after doing so.

The commit introduces the following:

**UI, API calls and view models to CRUD community permissions**

After creating a community, the user can go to the community settings and create new token permissions. The user can also update and delete existing permissions.

**Asset and collectible view models**

To create community token permissions, users have to select the token criteria. This commit introduces the `assetsModel` for ERC20 tokens and `collectiblesModel` for `ERC721` tokens.

The latter only supports custom minted community tokens at this point.

**This commit requires:** https://github.com/status-im/status-go/pull/3207

### CODE REVIEW

**Review by caybro — COMMENTED** [review]
Awesome job, just stumbled upon a couple of things

File: src/app/modules/main/chat_section/module.nim [hunk 1]
```diff
@@ -7,6 +7,11 @@ import model as chats_model
 import item as chat_item
 import ../../shared_models/user_item as user_item
 import ../../shared_models/user_model as user_model
+import ../communities/models/token_permissions_model
+import ../communities/models/token_permission_item
+import ../communities/models/token_criteria_item
+import ../communities/models/token_criteria_model
+import ../communities/models/token_list_item
```
Comment (jrainville) [inline src/app/modules/main/chat_section/module.nim:14]: Maybe those models and items should be put in `shared_models`?

File: src/app/modules/main/chat_section/module.nim [hunk 2]
```diff
@@ -307,6 +321,62 @@ proc initContactRequestsModel(self: Module) =
...
+  let erc20Tokens = self.controller.getTokenList()
+  let communityTokens = self.controller.getCommunityTokenList()
+
+  for token in erc20Tokens:
+    let tokenListItem = initTokenListItem(
+      token.symbol,
+      token.name,
+      token.symbol,
+      token.color,
+      "",
```
Comment (jrainville) [inline src/app/modules/main/chat_section/module.nim:357]: Could you add the arg name before for clarity? thanks

File: src/app/modules/main/chat_section/module.nim [hunk 3]
```diff
@@ -346,6 +416,8 @@ method load*(
     self.initContactRequestsModel()
   else:
     self.usersModule.load()
+    self.initCommunityTokenPermissionsModel()
+    self.buildTokenList()
```
Comment (jrainville) [inline src/app/modules/main/chat_section/module.nim:420]: I wonder if doing those on app start for each community might be slow? It doesn't have to be fixed now, but something to keep in mind. Maybe we'll want to lazy load the permissions or something

File: src/app/modules/main/chat_section/module.nim [hunk 4]
```diff
@@ -698,6 +770,27 @@ method onChatMuted*(self: Module, chatId: string) =
 method onChatUnmuted*(self: Module, chatId: string) =
   self.view.chatsModel().changeMutedOnItemById(chatId, muted=false)
 
+method onCommunityTokenPermissionDeleted*(self: Module, communityId: string, permissionId: string) =
+  self.view.tokenPermissionsModel().removeItemWithId(permissionId)
+  singletonInstance.globalEvents.showCommunityTokenPermissionDeletedNotification(communityId, "Community permission deleted", "A token permission has been removed")
+
+method onCommunityTokenPermissionCreated*(self: Module, communityId: string, tokenPermission: CommunityTokenPermissionDto) =
+  self.rebuildCommunityTokenPermissionsModel()
```
Comment (jrainville) [inline src/app/modules/main/chat_section/module.nim:778]: Can't we just add to the model? Wouldn't we only check if the new criteria is met anyway?

File: src/app/modules/main/chat_section/module.nim [hunk 5]
```diff
@@ -698,6 +770,27 @@ method onChatMuted*(self: Module, chatId: string) =
...
+  singletonInstance.globalEvents.showCommunityTokenPermissionCreatedNotification(communityId, "Community permission created", "A token permission has been added")
+
+method onCommunityTokenPermissionUpdated*(self: Module, communityId: string, tokenPermission: CommunityTokenPermissionDto) =
+  self.rebuildCommunityTokenPermissionsModel()
```
Comment (jrainville) [inline src/app/modules/main/chat_section/module.nim:782]: Same thing as above.

File: src/app_service/service/community/service.nim [hunk 6]
```diff
@@ -482,6 +506,51 @@ QtObject:
       self.allCommunities[community.id] = community
       self.events.emit(SIGNAL_COMMUNITIES_UPDATE, CommunitiesArgs(communities: @[community]))
 
+      # tokenPermission was added
+      if community.tokenPermissions.len > prev_community.tokenPermissions.len:
+        for id, tokenPermission in community.tokenPermissions:
+          if not prev_community.tokenPermissions.hasKey(id):
+            self.allCommunities[community.id].tokenPermissions[id] = tokenPermission
+
+            if not community.admin:
```
Comment (jrainville) [inline src/app_service/service/community/service.nim:515]: I don't think this condition makes sense. We'll have multiple admins in the future and you want to update the models when another admin added a permission.

IMO, the best way to handle it is to make sure that when we do `addPermission` (or whatever the name) below, we update `allCommunities`  communities accordingly, so the condition above `if community.tokenPermissions.len > prev_community.tokenPermissions.len:` doesn't even trigger for the current admin.

File: src/app_service/service/community/service.nim [hunk 7]
```diff
@@ -482,6 +506,51 @@ QtObject:
...
+              self.events.emit(SIGNAL_COMMUNITY_TOKEN_PERMISSION_CREATED,
+              CommunityTokenPermissionArgs(communityId: community.id, tokenPermission: tokenPermission))
+      elif community.tokenPermissions.len < prev_community.tokenPermissions.len:
+        for id, prvTokenPermission in prev_community.tokenPermissions:
+          if not community.tokenPermissions.hasKey(id):
+            if not community.admin:
```
Comment (jrainville) [inline src/app_service/service/community/service.nim:521]: Same thing

File: src/app_service/service/community/service.nim [hunk 8]
```diff
@@ -482,6 +506,51 @@ QtObject:
...
+            let index = findIndexBySymbol(tc.symbol, prevTokenPermission.tokenCriteria)
+            if index == -1:
+              continue
+
+            let prevTc = prevTokenPermission.tokenCriteria[index]
+            if tc.amount != prevTc.amount or tc.ensPattern != prevTc.ensPattern:
+              permissionUpdated = true
+              break
+
+          if permissionUpdated and not community.admin:
```
Comment (jrainville) [inline src/app_service/service/community/service.nim:550]: Same thing

File: src/app_service/service/community/service.nim [hunk 9]
```diff
@@ -563,6 +632,13 @@ QtObject:
   proc getCuratedCommunities*(self: Service): seq[CuratedCommunity] =
     return toSeq(self.curatedCommunities.values)
 
+  proc getCommunityByIdFromAllCommunities*(self: Service, communityId: string): CommunityDto =
```
Comment (jrainville) [inline src/app_service/service/community/service.nim:635]: Surprised we didn't have that function before. Oh well lol.

File: src/app_service/service/community/service.nim [hunk 10]
```diff
@@ -1528,3 +1604,53 @@ QtObject:
     except Exception as e:
       error "Error extracting discord channels and categories", msg = e.msg
 
+  proc editCommunityTokenPermission*(self: Service, communityId: string, tokenPermission: CommunityTokenPermissionDto) =
```
Comment (jrainville) [inline src/app_service/service/community/service.nim:1607]: Maybe `createOrEditCommunityTokenPermission` would be a clearer name.

File: src/app_service/service/community/service.nim [hunk 11]
```diff
@@ -1528,3 +1604,53 @@ QtObject:
...
+
+      if editing:
+        response = status_go.editCommunityTokenPermission(communityId, tokenPermission.id, int(tokenPermission.`type`), Json.encode(tokenPermission.tokenCriteria), tokenPermission.isPrivate)
+      else:
+        response = status_go.createCommunityTokenPermission(communityId, int(tokenPermission.`type`), Json.encode(tokenPermission.tokenCriteria), tokenPermission.isPrivate)
+
+      if response.result != nil and response.result.kind != JNull:
+        var changesField = "tokenPermissionsAdded"
+        if editing:
+          changesField = "tokenPermissionsModified"
```
Comment (jrainville) [inline src/app_service/service/community/service.nim:1621]: Maybe those should be in `const`s?

File: src/app_service/service/community/service.nim [hunk 12]
```diff
@@ -1528,3 +1604,53 @@ QtObject:
...
+
+        for permissionId, permission in response.result["communityChanges"].getElems()[0][changesField].pairs():
+          let p = permission.toCommunityTokenPermissionDto()
+          self.allCommunities[communityId].tokenPermissions[permissionId] = p
+          self.joinedCommunities[communityId].tokenPermissions[permissionId] = p
```
Comment (jrainville) [inline src/app_service/service/community/service.nim:1626]: Ah you actually already do the thing I commented above. The check for `community.admin` is then double useless :smile:

File: src/app_service/service/token/service.nim [hunk 13]
```diff
@@ -92,13 +96,49 @@ QtObject:
...
         )
+
+        let nativeToken = newTokenDto(
+          network.nativeCurrencyName,
+          network.chainId,
+          Address.fromHex("0x0000000000000000000000000000000000000000"),
+          network.nativeCurrencySymbol,
+          network.nativeCurrencyDecimals,
+          false,
+          ""
```
Comment (jrainville) [inline src/app_service/service/token/service.nim:107]: Can you add the name of the args for clarity? Thanks

File: src/app_service/service/wallet_account/dto.nim [hunk 14]
```diff
@@ -147,6 +147,13 @@ proc getAddress*(self: WalletTokenDto): string =
 
   return ""
 
+proc getTotalBalanceOfSupportedChains*(self: WalletTokenDto): float64 =
+  var sum = 0.0
+  for chainId, balanceDto in self.balancesPerChain:
+    sum += balanceDto.balance
```
Comment (jrainville) [inline src/app_service/service/wallet_account/dto.nim:153]: Should we use some some of bigInt for that? I'm worried it might overflow with the crazy decimals.

File: ui/app/AppLayouts/Chat/stores/CommunitiesStore.qml [hunk 15]
```diff
@@ -20,112 +21,72 @@ QtObject {
...
+    //                    },
+    //                    {
+    //                        key: "Custom",
+    //                        iconSource: "qrc:imports/assets/png/collectibles/SNT.png",
+    //                        name: "Custom Collectible",
+    //                        category: TokenCategories.Category.General
+    //                    }
+    //                ])
+    //     }
+    // }
```
Comment (jrainville) [inline ui/app/AppLayouts/Chat/stores/CommunitiesStore.qml:89]: Can we remove or we still uncomment from time to time?

**Review by jrainville — COMMENTED** [review]
**Massive** work dude! Good job!

I left a lot of comments, but that's because there are a lot of lines :laughing: 
None of them are super severe anyway.

Not a reproach,. but if you wanted, you could have just done `create` for now an added `edit` and `delete` later :smile:

File: src/app_service/service/community/dto/community.nim [hunk 16]
```diff
@@ -132,6 +161,62 @@ proc toDiscordImportTaskProgress*(jsonObj: JsonNode): DiscordImportTaskProgress
...
+  result = TokenCriteriaDto()
+  discard jsonObj.getProp("amount", result.amount)
+  discard jsonObj.getProp("decimals", result.decimals)
+  discard jsonObj.getProp("symbol", result.symbol)
+  discard jsonObj.getProp("name", result.name)
+  discard jsonObj.getProp("ensPattern", result.ensPattern)
+
+  var typeInt: int
+  discard jsonObj.getProp("type", typeInt)
+  if (typeInt >= ord(low(TokenCriteriaType)) or typeInt <= ord(high(TokenCriteriaType))):
```
[the remark this answers is not in the document]
Comment (0x-r4bbit) [reply src/app_service/service/community/dto/community.nim:174]: Ah, I used this check from another place where we do the same thing.
`and` would make indeed more sense. Good catch!

File: ui/app/AppLayouts/Chat/stores/CommunitiesStore.qml [hunk 17]
```diff
@@ -20,112 +21,72 @@ QtObject {
...
-                           key: "Amp",
-                           iconSource: "qrc:imports/assets/png/tokens/CUSTOM-TOKEN.png",
-                           name: "Amp",
-                           shortName: "AMP",
-                           category: TokenCategories.Category.Own
-                       }
-                   ])
-        }
-    }
+    property var assetsModel: chatCommunitySectionModule.tokenList
```
[the remark this answers is not in the document]
Comment (0x-r4bbit) [reply ui/app/AppLayouts/Chat/stores/CommunitiesStore.qml:24]: It's globally available like `communitiesModule` and `mainModule` (which are used in this store)

File: src/app/modules/main/chat_section/module.nim (hunk 1 above)
Comment (0x-r4bbit) [reply src/app/modules/main/chat_section/module.nim:14]: Can move them there for sure!
Started having them here because I believe we're only using them in the context of communities (and chat sections that **are** communities)

File: src/app/modules/main/chat_section/module.nim (hunk 2 above)
Comment (0x-r4bbit) [reply src/app/modules/main/chat_section/module.nim:357]: Will do!

File: src/app/modules/main/chat_section/module.nim (hunk 3 above)
Comment (0x-r4bbit) [reply src/app/modules/main/chat_section/module.nim:420]: Hm.. so this really just puts together some view models. I haven't measured it, but I'd be surprised if that would be a problem performance-wise (also given that you likely won't have hundreds of community instances

File: src/app/modules/main/chat_section/module.nim (hunk 4 above)
Comment (0x-r4bbit) [reply src/app/modules/main/chat_section/module.nim:778]: So originally I had it indeed implemented such that `onCommunityTokenPermissionCreated()` and `onCommunityTokenPermissionUpdated()` would just add and update an item respectively.

And because we need to revalidate whether we have still meet all token criteria when a permission was updated, I ended up using the rebuildCommunityTokenPermissionsModel() which does the revalidation but also updates the model.

However, now that I'm writing this, I realized there's a different issue with
`rebuildCommunityTokenPermissionsModel()` w.r.t. updateing the `allTokenCriteriaMet` flag...

I'll revisit this.

File: src/app_service/service/community/service.nim (hunk 6 above)
Comment (0x-r4bbit) [reply src/app_service/service/community/service.nim:515]: Okay so... the reason I'm checking for `community.admin` here is because, we're already emitting the signal (as admin) in the service after creation/update/deletion, so that would result in the signal being emitted twice, and twice the amount of app notifications etc.

If however `community.admin` is no longer == "owner", then this indeed needs to change.

Another alternative would be to not emit the signal from the service in the first place, and purely rely on `handleCommunityUpdates` to trigger the signal here. That means however, that signal has to come in first, before we see a change in the UI (which might not be a problem, if we know that signal will happen synchronously via MessengerResponse, after the create/update permission via service was done)

> IMO, the best way to handle it is to make sure that when we do addPermission (or whatever the name) below, we update allCommunities communities accordingly, so the condition above if community.tokenPermissions.len > prev_community.tokenPermissions.len: doesn't even trigger for the current admin.

Not super sure I follow. We already update `allCommunities` right now but we still want to know if there was a change in the tokenPermissions to emit a signal (cause this could be a normal member node)

File: src/app_service/service/community/service.nim (hunk 9 above)
Comment (0x-r4bbit) [reply src/app_service/service/community/service.nim:635]: We were relying on `joinedCommunities` when we did things like `getCommunityByID`.
Unfortunately that list of communities won't do the trick when dealing with `spectatedCommunities` (which is what we have when looking at a community that we haven't joined yet and might be token gated).

This has actually bitten be twice very badly.

@osmaczko created and issue about this a while ago: https://github.com/status-im/status-desktop/issues/9570

Bottom line is that we should just have `allCommunities`

File: src/app_service/service/community/service.nim (hunk 6 above)
Comment (jrainville) [reply src/app_service/service/community/service.nim:515]: > Another alternative would be to not emit the signal from the service in the first place, and purely rely on handleCommunityUpdates to trigger the signal here. That means however, that signal has to come in first, before we see a change in the UI (which might not be a problem, if we know that signal will happen synchronously via MessengerResponse, after the create/update permission via service was done)

This is not super reliable. There is often a delay between the signal comes back and when we do the action, so it creates a delay on the front-end.

File: src/app_service/service/wallet_account/dto.nim (hunk 14 above)
Comment (0x-r4bbit) [reply src/app_service/service/wallet_account/dto.nim:153]: Had a big discussion around that with @alaibe 
I think the biggest supported float is big enough to not worry to much about that here.
OTOH I don't think there's a theoretical decimals limit, so there could indeed be a token that goes bananas here.

It would require us to switch to `u256` all the way down tho. Which would also require a big refactor of the wallet if I'm not mistaken.

File: src/app_service/service/community/service.nim (hunk 6 above)
Comment (0x-r4bbit) [reply src/app_service/service/community/service.nim:515]: I believe the easiest path forward would be a distinction between `owner` and `admin`
Something that lets us express: "If you were the one who created/updated this permission, you don't need this signal"

Comment (jrainville) [reply src/app_service/service/community/service.nim:515]: > Not super sure I follow. We already update allCommunities right now but we still want to know if there was a change in the tokenPermissions to emit a signal (cause this could be a normal member node)

Yeah don't worry, you already do the necessary update to `allCommunities`, so that means that ` if not prev_community.tokenPermissions.hasKey(id)` would not even trigger, as far as I can tell. Therefore, I don't think you need  `if not community.admin:`.

File: src/app_service/service/community/service.nim (hunk 9 above)
Comment (caybro) [reply src/app_service/service/community/service.nim:635]: The real bottom line imho is that (and this pertains to all our NIM models) we should never do any sorting/filtering/bucketing on the backend side. Just like here, if a community has a `joined` model role, we can easily filter by that in QML

Comment (0x-r4bbit) [reply src/app_service/service/community/service.nim:635]: I believe the "bucketing" is a result of historic changes throughout the lifetime of the protocol.

But yes, we've already agreed in #9570 that we'll just have **one** list of known communities and then dedicated flags to know whether or not they are "joined" or "curated" or neither if necessary.

**Comment by 0x-r4bbit** [comment]
Apart from status-go conflicting, I"ve addressed all comments

**Review by jrainville — APPROVED** [review]
LGTM. Great job!

**Review by caybro — APPROVED** [review]
Code LGTM
